extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts

```diff
@@ -0,0 +1,204 @@
+				multiLocations: true,
+				defaultLocation: undefined
+			};
+		}
+
+		if (this._options.defaultLocation && !this._packageTypes.get(this._options.defaultLocation)) {
+			throw new Error(`Invalid default location '${this._options.defaultLocation}`);
+		}
+
+		if (!this._options.multiLocations && !this.defaultLocation) {
+			throw new Error('Default location not specified for single location mode');
+		}
+	}
+
+	/**
+	 * Returns the providers map
+	 */
+	public get packageManageProviders(): Map<string, IPackageManageProvider> {
+		return this._packageManageProviders;
+	}
+
+	/**
+	 * Returns the current provider
+	 */
+	public get currentPackageManageProvider(): IPackageManageProvider {
+		if (this.currentProvider) {
+			let provider = this._packageManageProviders.get(this.currentProvider);
+			return provider;
+		}
+		return undefined;
+	}
+
+	/**
+	 * Returns true if multi locations mode is enabled
+	 */
+	public get multiLocationMode(): boolean {
+		return this._options.multiLocations;
+	}
+
+	/**
+	 * returns the array of target locations
+	 */
+	public targetLocations(): string[] {
+		return Array.from(this._locations.keys());
+	}
+
+	/**
+	 * Returns the default location
+	 */
+	public get defaultLocation(): string {
+		return this._options.defaultLocation || this.targetLocations()[0];
+	}
+
+	/**
+	 * Loads the provider cache
+	 */
+	private async loadCaches(): Promise<void> {
+		if (this.packageManageProviders) {
+			let keyArray = Array.from(this.packageManageProviders.keys());
+			for (let index = 0; index < keyArray.length; index++) {
+				const element = this.packageManageProviders.get(keyArray[index]);
+				if (await element.canUseProvider()) {
+					if (!this._locations.has(element.packageTarget.location)) {
+						this._locations.add(element.packageTarget.location);
+					}
+					if (!this._packageTypes.has(element.packageTarget.location)) {
+						this._packageTypes.set(element.packageTarget.location, []);
+					}
+					this._packageTypes.get(element.packageTarget.location).push(element);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Returns the package types for given location
+	 */
+	public getPackageTypes(targetLocation?: string): [string, string][] {
+		targetLocation = targetLocation || this.defaultLocation;
+		let providers = this._packageTypes.get(targetLocation);
+		return providers.map(x => [x.providerId, x.packageTarget.packageType]);
+	}
+
+	/**
+	 * returns the list of packages for current provider
+	 */
+	public async listPackage(): Promise<IPackageDetails[]> {
+		let provider = this.currentPackageManageProvider;
+		if (provider) {
+			return await provider.listPackages();
+		} else {
+			return Promise.reject('Current Provider is not set');
+		}
+	}
+
+	/**
+	 * Changes the current provider
+	 */
+	public changeProvider(providerId: string): void {
+		if (this._packageManageProviders.has(providerId)) {
+			this.currentProvider = providerId;
+		} else {
+			throw Error(`Invalid package type ${providerId}`);
+		}
```

Comment:
One nit is that it's confusing switching between calling it "package type" and "provider ID". Why not just stick with one consistently throughout?

Reply:
fixed that. it should only change the provider which has the package type

---
In reply to: [350879426](https://github.com/microsoft/azuredatastudio/pull/8473#discussion_r350879426) [](ancestors = 350879426)